documentation/reference/ci_cd.md

```diff
@@ -0,0 +1,489 @@
+the `do_build` output of the [`determine-images`](#determine-images) job.
+
+## Catalog jobs
+
+### `test-cat`
+
+Runs tests for the catalog using the `catalog/test` recipe. Tests are run inside
+a Docker container so neither Python nor Node.js needs to be installed.
+
+This job is skipped if the catalog codebase has not changed. Its counterparts
+are [`test-ing`](#test-ing) for the ingestion server, [`test-api`](#test-api)
+for the API and [`nuxt-build`](#nuxt-build) for the frontend.
+
+```{note}
+This job is treated as the proof of functionality for publishing Docker images
+for the catalog.
+```
+
+### `catalog-checks`
+
+Runs tests for the catalog using the `catalog/generate-dag-docs` recipe. The job
+generates a new file from a Docker container so Python does not need to be
+installed. However, the job runs linting on the new file, so Node.js needs to be
+set up.
+
+This job is skipped if the catalog codebase has not changed.
+
+## Ingestion server jobs
+
+### `test-ing`
+
+Runs tests for the ingestion server using the `ingestion_server/test-local`
+recipe. Tests are run on the host so Python needs to be installed and set up.
+
+This job is skipped if the ingestion server codebase has not changed. Its
+counterparts are [`test-cat`](#test-cat) for the catalog,
+[`test-api`](#test-api) for the API and [`nuxt-build`](#nuxt-build) for the
+frontend.
+
+```{note}
+This job is treated as the proof of functionality for publishing Docker images
+for the ingestion server.
+```
+
+## API jobs
+
+### `test-api`
+
+Initialises the API using the `api/init` recipe and runs tests for the API using
+the `api/test` recipe. Tests are run inside a Docker container so neither Python
+nor Node.js needs to be installed.
+
+This job is skipped if the API codebase has not changed. Its counterparts are
+[`test-cat`](#test-cat) for the catalog, [`test-ing`](#test-ing) for the
+ingestion server and [`nuxt-build`](#nuxt-build) for the frontend.
+
+```{note}
+This job is treated as the proof of functionality for publishing Docker images
+for the API. Since it also initialises the API, it also largely proves that the
+ingestion server is working.
+```
+
+### `django-checks`
+
+Runs a matrix of various Django checks for the API using the following `just`
+recipes.
+
+- `api/dj check`
+- `api/dj validateopenapischema`
+- `api/dj makemigrations --check --noinput --merge`
+- `api/doc-test`
+
+This job is skipped if the API codebase has not changed. Its counterpart is
+[`nuxt-checks`](#nuxt-checks) for the frontend.
+
+#### `bypass-django-checks`
+
+The `django-checks` job is marked as required check in GitHub. If the job is
+skipped, the matrix will not be expanded into individual jobs and all required
+checks will appear as pending, thus blocking PRs from being merged.
+
+To counter this, the `bypass-django-checks` job is run if the API codebase has
+not changed, and it simply passes the checks by having the same job names.
```

Comment:
This blurb is shared by all of the `bypass-*` jobs and is very similar, does Sphinx provide any mechanism for us having a repeated section with slightly different references? Or perhaps, could we define the bypass info once and then reference that section in each of these places? Just so we don't need to update several places if this description were to be changed!

Reply:
I copy pasted that in three places because I assumed no one would read the entire file and the reader would most probably use the right sidebar to jump to the relevant section directly. That's also why this page has so many internal links between different jobs and actions.

Reply:
Could there be a "bypass jobs" section and each of the jobs that have it link to it? Something like:

```md
### Bypass jobs

Some jobs are marked as required check in GitHub. If the job is
skipped, the matrix will not be expanded into individual jobs and all required
checks will appear as pending, thus blocking PRs from being merged.

To counter this, those jobs have "bypass" jobs named in the pattern "bypass-{job name}".
```

````md
### `playwright`

Runs a matrix of various Playwright tests for the frontend using the following
Node.js scripts.

- `test:playwright visual-regression`
- `test:playwright e2e`
- `test:storybook`

This job is skipped if the frontend codebase has not changed.

```{note}
This job, combined with the [`nuxt-build` job](#nuxt-build), is treated as the
proof of functionality for deploying frontend to staging.
```

```{note}
[This job has an accompanying bypass job](#bypass-jobs).
```
````

Still some duplication in needing to paste the note everywhere, but at least the full text is deduplicated.

Reply:
The description of the bypass jobs needs more updating to make it more general, but hopefully the idea makes sense :sweat_smile:

Reply:
This makes sense, de-duplicated the text in c52f0bb.